stl/inc/stop_token

```diff
@@ -0,0 +1,401 @@
+            // another thread already requested
+            return false;
+        }
+
+        _Stopping_thread = _Thrd_id();
+        for (;;) {
+            auto _Head = _Callbacks._Lock_and_load();
+            _Current_callback.store(_Head, memory_order_relaxed);
+            _Current_callback.notify_one();
+            if (_Head == nullptr) {
+                _Callbacks._Store_and_unlock(nullptr);
+                return true;
+            }
+
+            const auto _Next = _STD exchange(_Head->_Next, nullptr);
+            _STL_INTERNAL_CHECK(_Head->_Prev == nullptr);
+            if (_Next != nullptr) {
+                _Next->_Prev = nullptr;
+            }
+
+            _Callbacks._Store_and_unlock(_Next); // unlock before running _Head so other registrations
+                                                 // can detach without blocking on the callback
+
+            _Head->_Fn(_Head); // might destroy *_Head
+        }
+    }
+};
+
+class stop_source;
+
+class stop_token {
+    friend stop_source;
+    friend _Stop_callback_base;
+
+public:
+    stop_token() noexcept : _State{} {}
+    stop_token(const stop_token& _Other) noexcept : _State{_Other._State} {
+        const auto _Local = _State;
+        if (_Local != nullptr) {
+            _Local->_Stop_tokens.fetch_add(1, memory_order_relaxed);
+        }
+    }
+
+    stop_token(stop_token&& _Other) noexcept : _State{_STD exchange(_Other._State, nullptr)} {}
+    stop_token& operator=(const stop_token& _Other) noexcept {
+        stop_token{_Other}.swap(*this);
+        return *this;
+    }
+
+    stop_token& operator=(stop_token&& _Other) noexcept {
+        stop_token{_STD move(_Other)}.swap(*this);
+        return *this;
+    }
+
+    ~stop_token() {
+        const auto _Local = _State;
+        if (_Local != nullptr) {
+            if (_Local->_Stop_tokens.fetch_sub(1, memory_order_acq_rel) == 1) {
+                delete _Local;
+            }
+        }
+    }
+
+    void swap(stop_token& _Other) noexcept {
+        _STD swap(_State, _Other._State);
+    }
+
+    [[nodiscard]] bool stop_requested() const noexcept {
+        const auto _Local = _State;
+        return _Local != nullptr && _Local->_Stop_requested();
+    }
+
+    [[nodiscard]] bool stop_possible() const noexcept {
+        const auto _Local = _State;
+        return _Local != nullptr && _Local->_Stop_possible();
+    }
+
+    [[nodiscard]] friend bool operator==(const stop_token& _Lhs, const stop_token& _Rhs) noexcept {
+        return _Lhs._State == _Rhs._State;
+    }
+
+    friend void swap(stop_token& _Lhs, stop_token& _Rhs) noexcept {
+        _STD swap(_Lhs._State, _Rhs._State);
+    }
+
+private:
+    explicit stop_token(_Stop_state* const _State_) : _State{_State_} {}
+
+    _Stop_state* _State;
+};
+
+class stop_source {
+public:
+    stop_source() : _State{new _Stop_state} {}
+    explicit stop_source(nostopstate_t) noexcept : _State{} {}

```

Comment:
Have you considered implementing nonstopable as a special state from a global variable for which ref count never reaches zero?  It may slightly optimize stoppables by pessimizing nonstopables.

Reply:
I don't think that buys us anything, because the `nostopstate_t` case is also the moved-from case. Plus it would force adding special cases to detect that special state because it would need to be `stop_possible`. Null checks are cheap :)